pkgs/_macro_host/lib/src/macro_host.dart

```diff
@@ -0,0 +1,188 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:_macro_builder/macro_builder.dart';
+import 'package:_macro_runner/macro_runner.dart';
+import 'package:_macro_server/macro_server.dart';
+import 'package:dart_model/dart_model.dart';
+import 'package:macro_service/macro_service.dart';
+
+import 'macro_cache.dart';
+import 'package_config.dart';
+
+/// Hosts macros: builds them, runs them, serves the macro service.
+///
+/// Tools that want to support macros, such as the Analyzer and the CFE, can
+/// do so by running a `MacroHost` and providing their own `HostService`.
+class MacroHost {
+  final MacroPackageConfig macroPackageConfig;
+  final _HostService _hostService;
+  final MacroServer macroServer;
+  final MacroBuilder macroBuilder = MacroBuilder();
+  final MacroRunner macroRunner = MacroRunner();
+  final MacroResultsCache _macroResultsCache;
+
+  /// We only want to expose the public interface and not the private impl.
+  HostService get hostService => _hostService;
+
+  /// Helper for internal use only.
+  QueryService get _queryService => _hostService.queryService;
+
+  MacroHost._(this.macroPackageConfig, this.macroServer, this._hostService)
+      : _macroResultsCache = MacroResultsCache(_hostService.queryService);
+
+  /// Starts a macro host with introspection queries handled by [queryService].
+  static Future<MacroHost> serve({
+    // TODO(davidmorgan): support serving multiple protocols.
+    required Protocol protocol,
+    required Uri packageConfig,
+    required QueryService queryService,
+  }) async {
+    final macroPackageConfig = MacroPackageConfig.readFromUri(packageConfig);
+    final hostService = _HostService(queryService);
+    final server =
+        await MacroServer.serve(protocol: protocol, service: hostService);
+    return MacroHost._(macroPackageConfig, server, hostService);
+  }
+
+  /// Whether [name] is a macro according to that package's `pubspec.yaml`.
+  bool isMacro(QualifiedName name) => lookupMacroImplementation(name) != null;
+
+  /// Checks whether [annotation] is a macro annotation.
+  ///
+  /// If so, returns the qualified name of the macro implementation.
+  ///
+  /// If not, returns `null`.
+  QualifiedName? lookupMacroImplementation(QualifiedName annotation) =>
+      macroPackageConfig.lookupMacroImplementation(annotation);
+
+  /// Determines which phases the macro triggered by [annotation] runs in.
+  Future<Set<int>> queryMacroPhases(
+      Uri packageConfig, QualifiedName annotation) async {
+    await _ensureRunning(annotation);
+    return _hostService._macroState[annotation.asString]!.phases;
+  }
+
+  /// Sends [request] to the macro triggered by [annotation].
+  Future<AugmentResponse> augment(
+      QualifiedName annotation, AugmentRequest request) async {
+    final cached = await _macroResultsCache.cachedResult(annotation, request);
```

Comment:
It took me a while to figure out what this is doing: rerun the queries from last time, discard the output from the cache if anything changed.

It looks like it would make sense to also cache the query results here, so that if we do run the macro then the ones we fetched to check invalidation can be used instead of requested again?

I guess a TODO for that would make sense :)

Reply:
The queries themselves could change too, but it might make sense to try something like that